pkg/api/sidecar/v1.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package v1
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+
+	"github.com/go-kit/log"
+	"github.com/opentracing/opentracing-go"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/route"
+	"github.com/thanos-io/thanos/pkg/promclient"
+	"github.com/thanos-io/thanos/pkg/shipper"
+
+	"github.com/thanos-io/thanos/pkg/api"
+	extpromhttp "github.com/thanos-io/thanos/pkg/extprom/http"
+	"github.com/thanos-io/thanos/pkg/logging"
+)
+
+// SidecarAPI is a very simple API used by Thanos Sidecar.
+type SidecarAPI struct {
+	baseAPI *api.BaseAPI
+	client  *promclient.Client
+	shipper *shipper.Shipper
+	promURL *url.URL
+	dataDir string
+	logger  log.Logger
+	reg     prometheus.Registerer
+}
+
+// NewSidecarAPI creates an Thanos Sidecar API.
+func NewSidecarAPI(
+	logger log.Logger,
+	reg prometheus.Registerer,
+	client *promclient.Client,
+	shipper *shipper.Shipper,
+	dataDir string,
+	promURL *url.URL,
+) *SidecarAPI {
+	return &SidecarAPI{
+		baseAPI: api.NewBaseAPI(logger, true, nil),
+		logger:  logger,
+		client:  client,
+		reg:     reg,
+		shipper: shipper,
+		dataDir: dataDir,
+		promURL: promURL,
+	}
+}
+
+func (s *SidecarAPI) Register(r *route.Router, tracer opentracing.Tracer, logger log.Logger, ins extpromhttp.InstrumentationMiddleware, logMiddleware *logging.HTTPServerMiddleware) {
+	s.baseAPI.Register(r, tracer, logger, ins, logMiddleware)
+
+	instr := api.GetInstr(tracer, logger, ins, logMiddleware, true)
+	r.Post("/flush", instr("flush", s.flush))
+}
+
+type flushResponse struct {
+	BlocksUploaded int `json:"blocksUploaded"`
+}
+
+func (s *SidecarAPI) flush(r *http.Request) (interface{}, []error, *api.ApiError, func()) {
+	dir, err := s.client.Snapshot(r.Context(), s.promURL, false)
+
+	if err != nil {
+		return nil, nil, &api.ApiError{Typ: api.ErrorInternal, Err: fmt.Errorf("failed to snapshot: %w", err)}, func() {}
+	}
+
+	snapshotDir := s.dataDir + "/" + dir
+
+	s.shipper.SetDirectoryToSync(snapshotDir)
+	uploaded, err := s.shipper.Sync(r.Context())
```

Comment:
Prometheus might produce a block in the middle of the snapshot call. I think you will have to disable the other syncer before doing anything.

Reply:
Is the concern that it'd produce an overlapping block? Or that there's a potential race between two shippers `Sync`ing at the same time?